startup.s

```diff
@@ -51,31 +52,46 @@ _vectors:
 
 Reset_Handler:
 
+#ifdef USE_RISKY_SHORTCUT
+  /* Address of load_addresses, determined by experimentation because linker
+   * doesn't support symbols for thumb immediates. Although unlikely, it's
+   * possible load_addresses could end up at a different location when
+   * using a different toolchain; hence this shortcut is disabled by default.
+   * When used, it saves 4 bytes. To enable it:
+   *
+   * DEFINES=-DUSE_RISKY_SHORTCUT make
+   *
+   * Then check that the address is actually correct before flashing it to
+   * your device.
+   */
+  movs r7, #0x40
+#else
+  ldr r7, =load_addresses
+#endif
+  ldm r7!, {r0, r1, r2, r3, r4}
+
   /* optimized to assume non-zero sized 'data' section */
-  ldr r0, =__data_end__
-  ldr r1, =__data_start__
-  ldr r2, =__data_load_start__
 copy_loop:
-  ldm r2!, {r3}
-  stm r1!, {r3}
+  ldm r2!, {r7}
+  stm r1!, {r7}
   cmp r1, r0
   bcc copy_loop
 
-  /* optimized to assume non-zero sized 'bss' section */
-  ldr r0, =__bss_end__
-  ldr r1, =__bss_start__
+  /* optimized to assume non-zero sized 'bss' section and that
+   * __bss_start__ equals __data_end__.
+   */
 zero_loop:
   movs r2, #0
   stm r1!, {r2}
-  cmp r1, r0
+  cmp r1, r3
   bcc zero_loop
 
+  /* Only r4-r8 saved by callee */
   bl bootloader
 
   /* if bootloader returns, we proceed to the user app (origin address now in r0) */
 
-  ldr r1, =0xE000ED08 /* VTOR register */
-  str r0, [r1] /* point VTOR to user app */
+  str r5, [r4] /* point VTOR to user app */

```

Comment:
This doesn't look right to me

Reply:
Indeed, the VTOR is zero which will break any application that uses interrupts. @majbthrd please give me a chance to review the changes to my PR before merging it next time.